draft-ietf-ppm-dap.md

```diff
@@ -1514,217 +1609,235 @@ to the next phase of the aggregation protocol.
-the new `prep_msg` message to the Helper in a POST request to the aggregation
-job URI used during initialization (see {{leader-init}}). The body of the
-request is an `AggregationJobContinueReq`:
+Next, the Leader sends a POST request to the aggregation job URI used during
+initialization (see {{leader-init}}) with media type
+"application/dap-aggregation-job-continue-req" and body structured as:
 
 ~~~
 struct {
-  u16 round;
+  u16 step;
   PrepareStep prepare_steps<1..2^32-1>;
 } AggregationJobContinueReq;
 ~~~
 
-The `round` field is the round of VDAF preparation that the Leader just reached
-and wants the Helper to advance to.
+The `step` field is the step of DAP aggregation that the Leader just reached
+and wants the Helper to advance to. The `prepare_steps` field is the sequence
+of prep steps constructed in the previous step. The `PrepareStep`s MUST be in
+the same order as the previous aggregate request.
 
-The `prepare_steps` field MUST be a sequence of `PrepareStep`s in the
-`continued` state containing the corresponding `inbound` prepare message.
+The Helper's response will be an `AggregationJobResp` message (see
+{{aggregation-helper-init}}), which the Leader validates according to the
+criteria in {{aggregation-job-validation}}. If the message is invalid, the
+Leader MUST abort the aggregation job.
 
-The media type is set to "application/dap-aggregation-job-continue-req".
+Otherwise, the Leader proceeds as follows with each report:
 
-The Helper's response will be an `AggregationJobResp` message (see
-{{aggregation-helper-init}}), which the Leader validates according to
-the criteria in {{aggregation-job-validation}}. If the message is valid, the
-Leader moves to the next round of continuation with the enclosed prepare steps.
-Otherwise, the Leader should abandon the aggregation job entirely.
+1. If the inbound prep step type is "continue" and the state is
+   `Continued(prep_state)`, then the Leader computes
+
+   ~~~
+   (state, outbound) = VDAF.ping_pong_req(agg_param, state, inbound)
+   ~~~
+
+   where `inbound` is the message payload. If `outbound != None`, then the
+   Leader stores `state` and `outbound` and proceeds to another continuation
+   step. If `outbound == None`, then the preparation process is complete: either
+   `state == Rejected()`, in which case the Leader rejects the report and
+   removes it from the candidate set; or `state == Finished(out_share)`, in
+   which case preparation is complete and the Leader stores the output share for
+   use in the collection sub-protocol {{collect-flow}}.
+
+1. Else if the type is "finished" and `state == Finished(out_share)`, then
+   preparation is complete and the Leader stores the output share for use in
+   the collection flow ({{collect-flow}}).
+
+1. Else if the type is "reject", then the Leader rejects the report and removes
+   it from the candidate set.
+
+1. Else the type is invalid, in which case the Leader MUST abort the
+   aggregation job.
```

Comment:
If we take my other suggestion to change the definition of `AggregationJobInitReq` (https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/393/files#r1228639257), then I think we could change the definition of `struct PrepareStep` such that invalid types are impossible, and then this is one more check implementations don't have to worry about.

Orthogonally to that suggestion, I think this text could be moved into `{{aggregation-job-validation}}`, just to avoid repeating it here and in the aggregation init section.

Reply:
(waiting to resolve the step-specific-message issue before addressing the first part of this comment)

I disagree with the second part -- if anything, I'd say we should "inline" the content of `{{aggregation-job-validation}]` and then eliminate that section entirely. The reasoning for this is that documents should be optimized for the reader over the writer/maintainer (IMO), readers have a much easier time comprehending a document that is readable in a linear fashion, and this particular section contains a single small validation check. Repetition is IMO justified in this case, as it leads to a more understandable document with very little cost to maintainability.

Also, note that the `Leader Initialization` & `Leader Continuation` next-step logic is slightly different, i.e. in handling of a `finished` incoming message. We could call out the special case in the section we factored this out to, but if we have to have a special case for every place we are referring to the section, IMO that makes the proposed factoring less valuable.

Reply:
Given that there's only a single item in `{{aggregation-job-validation}}`, I'm inclined to agree that we could inline that section. I think we could punt that to a later change since (IIRC) this PR does not modify `{{aggregation-job-validation}}`.

Reply:
This text is still needed: There is the case that the Leader is in state `Finished` but the inbound message is `continue`.

Reply:
Filed https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/473 so we don't have to do that change here.